site/src/pages/WorkspacePage/WorkspaceActions/DownloadLogsDialog.tsx

```diff
@@ -0,0 +1,166 @@
+import { saveAs } from "file-saver";
+import JSZip from "jszip";
+import { useMemo, useState, type FC } from "react";
+import { useQueries, useQuery } from "react-query";
+import { agentLogs, buildLogs } from "api/queries/workspaces";
+import type { Workspace, WorkspaceAgent } from "api/typesGenerated";
+import {
+  ConfirmDialog,
+  type ConfirmDialogProps,
+} from "components/Dialogs/ConfirmDialog/ConfirmDialog";
+import { displayError } from "components/GlobalSnackbar/utils";
+import { Stack } from "components/Stack/Stack";
+
+type DownloadLogsDialogProps = Pick<
+  ConfirmDialogProps,
+  "onConfirm" | "onClose" | "open"
+> & {
+  workspace: Workspace;
+  download?: (zip: Blob, filename: string) => void;
+};
+
+type DownloadableFile = {
+  name: string;
+  blob: Blob | undefined;
+};
+
+export const DownloadLogsDialog: FC<DownloadLogsDialogProps> = ({
+  workspace,
+  download = saveAs,
+  ...dialogProps
+}) => {
+  const theme = useTheme();
+  const agents = selectAgents(workspace);
+  const agentLogResults = useQueries({
+    queries: agents.map((a) => ({
+      ...agentLogs(workspace.id, a.id),
+      enabled: dialogProps.open,
+    })),
+  });
+  const buildLogsQuery = useQuery({
+    ...buildLogs(workspace),
+    enabled: dialogProps.open,
+  });
+  const downloadableFiles: DownloadableFile[] = useMemo(() => {
+    const files: DownloadableFile[] = [
+      {
+        name: `${workspace.name}-build-logs.txt`,
+        blob: buildLogsQuery.data
+          ? new Blob([buildLogsQuery.data.map((l) => l.output).join("\n")], {
+              type: "text/plain",
+            })
+          : undefined,
+      },
+    ];
+
+    agents.forEach((a, i) => {
+      const name = `${a.name}-logs.txt`;
+      const logs = agentLogResults[i].data;
+      const txt = logs?.map((l) => l.output).join("\n");
+      let blob: Blob | undefined;
+      if (txt) {
+        blob = new Blob([txt], { type: "text/plain" });
+      }
+      files.push({ name, blob });
+    });
+
+    return files;
+  }, [agentLogResults, agents, buildLogsQuery.data, workspace.name]);
+  const isLoadingFiles = downloadableFiles.some((f) => f.blob === undefined);
+  const [isDownloading, setIsDownloading] = useState(false);
+
+  return (
+    <ConfirmDialog
+      {...dialogProps}
+      hideCancel={false}
+      title="Download logs"
+      confirmText="Download"
+      disabled={isLoadingFiles}
+      confirmLoading={isDownloading}
+      onConfirm={async () => {
+        try {
+          setIsDownloading(true);
+          const zip = new JSZip();
+          downloadableFiles.forEach((f) => {
+            if (f.blob) {
+              zip.file(f.name, f.blob);
+            }
+          });
+          const content = await zip.generateAsync({ type: "blob" });
+          download(content, `${workspace.name}-logs.zip`);
+          dialogProps.onClose();
+          setTimeout(() => {
+            setIsDownloading(false);
+          }, theme.transitions.duration.leavingScreen);
```

Comment:
Just curious, why do we need this to be in a timeout?

Reply:
To not reset the modal state while it is fading out. It looks buggy when it happens while the out transition is happening.

Reply:
Ohhh that makes sense!